src/datautils/wellboreItemShapeGenerator.ts

```diff
@@ -1,78 +1,116 @@
-  const res = casings.filter((c: any) => c.casingId === id);
-  return res.length > 0 ? res[0] : {};
+export const parseCement = (cement: Cement, casings: Casing[], holes: HoleSize[]): CompiledCement => {
+  const attachedCasing = findCasing(cement.casingId, casings);
+  const res: CompiledCement = {
+    ...cement,
+    boc: attachedCasing.end,
+    attachedCasing,
+    intersectingItems: findIntersectingItems(cement, attachedCasing, casings, holes),
+  };
+  return res;
 };
 
-export const overlaps = (top1: number, bottom1: number, top2: number, bottom2: number): boolean => top1 <= bottom2 && top2 <= bottom1;
+export const cementDiameterChangeDepths = (
+  cement: CompiledCement,
+  diameterIntervals: {
+    start: number;
+    end: number;
+  }[],
+): number[] => {
+  const diameterChangeDepths =
+    merge(
+      diameterIntervals.map((d) => [
+        d.start - 0.0001, // +- 0.0001 to find diameter right beforeobject
+        d.start,
+        d.end,
+        d.end + 0.0001, // +- 0.0001 to find diameter right after object
+      ]),
+    ).filter((d) => d >= cement.toc && d <= cement.boc) as number[]; // trim
 
-export const findIntersectingItems = (cement: Cement, parentCasing: Casing, casings: Casing[], holes: HoleSize[]) => {
-  const { toc: start } = cement;
+  diameterChangeDepths.push(cement.toc);
+  diameterChangeDepths.push(cement.boc);
 
-  const res = [];
-  res.push(...holes.filter((h: HoleSize) => overlaps(start, parentCasing.end, h.start, h.end) && h.diameter > parentCasing.diameter));
-  res.push(
-    ...casings.filter(
-      (c: Casing) => c.casingId !== cement.casingId && overlaps(start, parentCasing.end, c.start, c.end) && c.diameter > parentCasing.diameter,
-    ),
-  );
-  return res;
+  const uniqDepths = uniq(diameterChangeDepths);
+
+  return uniqDepths.sort((a: number, b: number) => a - b);
+};
+
+export const calculateCementDiameter = (innerCasing: Casing, nonAttachedCasings: Casing[], holes: HoleSize[]) => (
+  depth: number,
+): {
+  md: number;
+  innerDiameter: number;
+  outerDiameter: number;
+} => {
+  const defaultCementWidth = 100; // Default to flow cement outside to show error in data
+
+  const innerDiameter = innerCasing ? innerCasing.diameter : 0;
+
+  const outerCasings = nonAttachedCasings.filter((casing) => casing.innerDiameter > innerDiameter);
+  const holeAtDepth = holes.find((casing) => casing.start <= depth && casing.end >= depth);
+
+  const outerObjectAtDepth = [...outerCasings, holeAtDepth]
+    .filter((d) => d)
+    .sort((a, b) => {
+      const aDim = a.innerDiameter ? a.innerDiameter : a.diameter;
+      const bDim = b.innerDiameter ? b.innerDiameter : b.diameter;
+      return aDim - bDim;
+    }) // ascending
+    .find((object) => object.start <= depth && object.end >= depth);
+
+  const outerDiameter = outerObjectAtDepth
+    ? outerObjectAtDepth.innerDiameter
+      ? outerObjectAtDepth.innerDiameter
+      : outerObjectAtDepth.diameter - HOLE_OUTLINE
+    : defaultCementWidth;
```

Comment:
I assume the check for `innerDiameter` on line 106 is used to distinguish between holes and casings. Is there any other method to check for this, that is possibly more self explaining? If not, then let's just add a comment about that